daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/layers/api/Layer.scala

```diff
@@ -0,0 +1,275 @@
+  prefix: String,
+  namespace: String,
+  variables: java.util.List[(String, DFDLPrimType)],
+)
+
+/**
+ * This is the primary API class for writing layers.
+ *
+ * All layers are derived from this class, and must have no-args default constructors.
+ *
+ * Derived classes will be dynamically loaded by Java's SPI system.
+ * The names of concrete classes derived from Layer are listed in a resources/M.services file
+ * so that they can be found and dynamically loaded.
+ *
+ * The SPI creates an instance the class of which is used as a factory to create the
+ * instances actually used by the Daffodil runtime. Compilation of the static information about
+ * the layer occurs only once and is then shared by all runtime instances.
+ *
+ * Instances of derived layer classes can be stateful. They are private to threads, and each time a layer
+ * is encountered during parse/unparse, an instance is created for that situation.
+ *
+ * All the static information about the layer is provided in the arguments.
+ *
+ * The rest of the Layer class implements the
+ * layer decode/encode logic, which is done as part of deriving one's Layer class from the
+ * Layer base class.
+ *
+ * About variables: Layer logic may read and write variables. Variables being read are parameters to
+ * the layer algorithm. Variables being written are outputs (such as checksums) from the layer algorithm.
+ * Variables being written must be undefined, since variables in DFDL are single-assignment.
+ * Variables being read must be defined before being read by the layer, and this is true for both
+ * parsing and unparsing. When unparsing, variables being read cannot be forward-referencing to parts
+ * of the DFDL infoset that have not yet been unparsed.
+ *
+ * @param layerName the name that will appear in the DFDL schema to identify the layer
+ * @param supportedLayerLengthKinds list of the layer length kinds the layer supports
+ * @param supportedLayerLengthUnits list of the layer length units the layer supports
+ * @param isRequiredLayerEncoding true if the layer is textual and so needs the layerEncoding property
+ * @param optLayerVariables a LayerVariables structure describing the variables the layer accesses
+ */
+abstract class Layer(
+  val layerName: String,
+  val supportedLayerLengthKinds: java.util.List[JLayerLengthKind],
+  val supportedLayerLengthUnits: java.util.List[JLayerLengthUnits],
+  val isRequiredLayerEncoding: Boolean,
+  val optLayerVariables: Optional[LayerVariables],
+) {
+
+  final def name(): String =
+    layerName // name() method with empty args is required by SPI loader
+
+  /**
+   * Called exactly once when the schema is compiled to do extra checking that the layer is being used properly.
+   * The thrown exception becomes a SchemaDefinitionError at schema compile time.
+   *
+   * Example checks are:
+   * - layerEncoding is constant and is a single-byte charset
+   * - layerLength, if constant, is within a maximum value range
```

Comment:
Another thought, can layerLengthKind="explicit" and layerLength be removed, and just use a parent element with dfdl:length

For example, what is the difference between these two:

```xml
<sequence dfdlx:layer="foo" dfdlx:layerLengthKind="explicit" dfdl:lengthLength="100">
  ...
</sequence>
```

```xml
<element name="foo" dfdl:lengthKind="explicit" dfdl:length="100">
  <complexType>
    <sequence dfdlx:layer="foo">
      ...
    </sequence>
  </complexType>
</element>
```

The latter is more verbose, but I think it acheives the same thing? The idea being layers don't have support a specific kind of layer length. A layer just always reads until it hits the end of data, which might be EOF or might be a length provided by a parent element, or might be restricted by a surrounding layer that limits the length.

Reply:
Yes, but there is a difference in behavior. 

The layerLengthKind 'explicit' is only explicit on parse. On unparse it ignores the layerLength. 

This was to avoid the problem DFDL v1.0 has where the output wants to be perhaps a different length than the input, but to recompute that on output you have to measure the length not of the original element, but an inner element inside it, to avoid the circularity. 

But I'm inclined to try getting rid of dfdl:layerLengthKind entirely and see how it goes.